File: orderit.rb
n = gets.to_i
gets
shuffled = Array.new(n) { gets.chomp }
gets
original = Array.new(n) { gets.chomp }
dp = Array.new(n + 1) { Array.new(n + 1, 0) }
(1..n).each do |i|
  (1..n).each do |j|
    if shuffled[i - 1] == original[j - 1]
      dp[i][j] = dp[i - 1][j - 1] + 1
    else
      dp[i][j] = [dp[i - 1][j], dp[i][j - 1]].max
    end
  end
end
puts n - dp[n][n]

File: orderit.php
<?php
$n = intval(trim(fgets(STDIN)));
fgets(STDIN);
$shuffled = [];
for ($i = 0; $i < $n; $i++) $shuffled[] = trim(fgets(STDIN));
fgets(STDIN);
$original = [];
for ($i = 0; $i < $n; $i++) $original[] = trim(fgets(STDIN));
$dp = array_fill(0, $n + 1, array_fill(0, $n + 1, 0));
for ($i = 1; $i <= $n; $i++) {
  for ($j = 1; $j <= $n; $j++) {
    if ($shuffled[$i - 1] === $original[$j - 1])
      $dp[$i][$j] = $dp[$i - 1][$j - 1] + 1;
    else
      $dp[$i][$j] = max($dp[$i - 1][$j], $dp[$i][$j - 1]);
  }
}
echo $n - $dp[$n][$n] . "\n";
?>
